src/computability/encoding.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2020 Pim Spelier, Daan van Gent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pim Spelier, Daan van Gent.
+-/
+
+import data.fintype.basic
+import data.num.lemmas
+import tactic
+
+/-!
+# Encodings
+
+This file contains the definition of a (finite) encoding, a map from a type to
+strings in an alphabet, used in defining computability by Turing machines.
+It also contains several examples:
+
+## Examples
+
+- `fin_encoding_nat_Γ₀₁`   : a binary encoding of ℕ in a simple alphabet.
+- `fin_encoding_nat_Γ'`    : a binary encoding of ℕ in the alphabet used for TM's.
+- `unary_fin_encoding_nat` : a unary encoding of ℕ
+- `fin_encoding_bool_Γ₀₁`  : an encoding of bool.
+-/
+
+namespace computability
+
+/-- An encoding of a type in a certain alphabet, together with a decoding. -/
+structure encoding (α : Type) :=
+(Γ : Type)
```

Comment:
My gut instinct is that the type variable should be a parameter here (possibly an `out_param` if you make it a class). This will keep everything in `Type`, and it will also play a bit better with the TM classes that use external type variables as well.

Reply:
I think we're keeping it a structure instead of a class, as there can be multiple interesting encodings on a type (even when the alphabet is fixed).

I don't know what "external type variables" are exactly? I don't know if this is relevant, but we ended up defining a structure fin_tm2, that bundles the alphabet, the function labels, states, functions, and some finiteness assumptions. (That'll be in the next PR, probably coming sometime soon.)

Reply:
I mean a parameter, like alpha in this definition.

Reply:
Then I think that, as we bundled it into the definition of a TM (in the upcoming PR), we'll also keep it bundled into `encoding` for now. But thanks, I'll make sure to keep an eye out for if it ever becomes a problem!